File: 64-minimum-path-sum/64-minimum-path-sum.py
class Solution:
    def minPathSum(self, grid: List[List[int]]) -> int:
        @cache
        def dp(i, j):
            if i == 0 and j == 0:
                return grid[i][j]
            
            if i >= len(grid) or i < 0 or j >= len(grid[0]) or j < 0:
                return float('inf')
            
            ans = grid[i][j] + min(dp(i -1, j), dp(i , j - 1))
            return ans
            
        return dp(len(grid) - 1, len(grid[0]) - 1)
            
                
            
        

File: 64-minimum-path-sum/64-minimum-path-sum.java
class Solution {
    public int minPathSum(int[][] grid) {
        int n = grid.length, m = grid[0].length;
        int [] memo = new int[m];
        for (int i = 0; i < n; ++i){
            for (int j = 0; j < m; ++j){
                if (i == 0 && j == 0){
                    memo[j] = grid[i][j];
                }
                else if (i == 0) {
                    memo[j] = grid[i][j] + memo[j - 1];
                }
               else if (j == 0){
                    memo[j] = grid[i][j] + memo[j];
                }
                else{
                    memo[j] = grid[i][j] + Math.min(memo[j], memo[j - 1]);
                }
            }
        }        
        return memo[m - 1];
    }
}